### PR TITLE
feat: 106 slot unit derivation

### LINKED ISSUE #106: Add a new getter on slots `derivedUnit` to fetch an associated unit if there one
We want to allow certain slots to have the unit `"any"` to allow creating generic slots for generic controller. a PID controller will no know what kind of units will be used for the sensor + target. "any" should allow connecting any unit to the slot 

- [ ] update the assertion for matching units when connecting slots to allow any connection, when one of the slots has a unit of `"any"`

In order to be able to properly display in the UI which unit a "target" expects, the target slot of a controller has to use the unit of the connected sensor. So we need a unitResolver within every functionality that returns you the desired unit, if it can be resovled. 
The basic logic is that: 

```
Slot.#derivedUnit() {
  return this.unit === 'any'
    ? this.functionality.interSlotUnitResolver(this)
    : this.unit
}

InSlot.#derivedUnit() { 
  const derivedSlot = super.derivedUnit()
  return derivedSlot || (this.sourceSlot && this.sourceSlot.derivedUnit) || this.unit
}

OutSlot.#derivedUnit() { 
  const derivedSlot = super.derivedUnit()
  return derivedSlot || this.unit
}
```

### PR DESCRIPTION
**Pull request by vdiezel**
closes #106

the first part for looking up the derived unit of a slot

For the exact mapping on special functionalities like Controllers, we need a list of slot names from the firmware

### CODE REVIEW

File: src/slots/Slot.js [hunk 1]
```diff
@@ -13,30 +13,10 @@ class Slot {
...
-  }
+  static get ANY_UNIT_STRING() { return 'any' }
 
-  static get UNITLESS_UNIT() { return Slot.UNIT_TYPE_UNIT_OPTIONS[Slot.UNIT_TYPES.UNITLESS][0] }
+  static get UNITLESS_UNIT() { return '-' }
 
-  static get ALL_UNIT_VALUES() {
-    return Object.values(this.UNIT_TYPE_UNIT_OPTIONS)
-      .reduce((acc, opts) => ([ ...acc, ...opts ]), [])
-  }
```
Comment (DanielSeehausen) [inline src/slots/Slot.js:39]: per discussion - request to keep this in. I will make a ticket to yank out the unit registry from here so we have a central place it is kept track of.

File: src/slots/Slot.js [hunk 2]
```diff
@@ -94,7 +76,7 @@ class Slot {
       name: Joi.string().required(),
       displayType: Joi.string().allow(null).required(),
       dataStreams: Joi.array().items(DataStream.SCHEMA).required(),
-      unit: Joi.string().allow(...this.ALL_UNIT_VALUES).required(),
```
Comment (DanielSeehausen) [inline src/slots/Slot.js:97]: could either leave in so when we add the specific unit validations its straightforward or yank for now understanding its not doing anything

File: test/functionalities/Functionality.test.js [hunk 3]
```diff
@@ -283,7 +284,67 @@ describe('the Functionality class', () => {
...
+            const controller = PIDControllerSeeder.seedOne({ slots: [ slot ] })
+
+            const slotWithinFct = controller.getSlotByName(slot.name)
+
+            expect(slotWithinFct.derivedUnit).toBe(unit)
+          })
+        })
+
+        describe('when it has a connection', () => {
+          it('returns the unit of its source', () => {
```
Comment (DanielSeehausen) [inline test/functionalities/Functionality.test.js:317]:
```suggestion
          it('returns the derived unit of its source', () => {
```
(assuming this updated test description is accurate?)

File: test/functionalities/Functionality.test.js [hunk 4]
```diff
@@ -283,7 +284,67 @@ describe('the Functionality class', () => {
...
 
+            const controller = PIDControllerSeeder.seedOne({ slots: [ slot ] })
+            const sensor = TemperatureSensorSeeder.seedOne({ slots: [ sensorSlot ] })
+
+            const controllerSlotWithinFct = controller.getSlotByName(slot.name)
+            const sensorSlotWithinFct = sensor.getSlotByName(sensorSlot.name)
+
+            sensorSlotWithinFct.connectTo(controllerSlotWithinFct)
+
+            expect(controllerSlotWithinFct.derivedUnit).toBe(sensorSlotUnit)
```
Comment (DanielSeehausen) [inline test/functionalities/Functionality.test.js:331]:
```suggestion
            expect(controllerSlotWithinFct.derivedUnit).toBe(sensor.derivedUnit)
```

**Review by DanielSeehausen — CHANGES_REQUESTED** [review]
only blocking request is to use the `derivedUnit` of the fallback slot instead of the `unit`, since they could return different things!

File: test/functionalities/Functionality.test.js (hunk 3 above)
Comment (vdiezel) [reply test/functionalities/Functionality.test.js:317]: fixed

File: test/functionalities/Functionality.test.js (hunk 4 above)
Comment (vdiezel) [reply test/functionalities/Functionality.test.js:331]: fixed

**Review by DanielSeehausen — APPROVED** [review]

**Pull request merged**